scraping_utils.py

```diff
@@ -0,0 +1,305 @@
+
+#############################################################################
+# TIMETABLE SCRAPING
+#############################################################################
+
+# Querying Timetable
+def get_hidden_inputs(text):
+    return BeautifulSoup(text, 'html.parser').find_all("input", type="hidden")
+
+def update_form(old_form, new_form):
+        new_form = {x["id"]:x["value"] for x in new_form}
+        old_form.update({x:y for x, y in new_form.items() if y.strip() != ''})
+        old_form['ICAction'] = 'CLASS_SRCH_WRK2_SSR_PB_CLASS_SRCH'
+
+def _separate_requests_query():
+    r = requests.get(orig_link)
+    new_form = get_hidden_inputs(r.text)
+    update_form(old_form, new_form)
+    r = requests.post(orig_link, data=old_form, headers=default_headers,
+                      cookies=r.cookies)
+    return r
+
+def format_query(year, term, subject, number):
+    query = old_form
+    term = term.lower()
+    query["CLASS_SRCH_WRK2_STRM$35$"] = "2" + str(year)[-2:] + term_to_num[term]
+    query["SSR_CLSRCH_WRK_SUBJECT$0"] = subject.lower()
+    query["SSR_CLSRCH_WRK_CATALOG_NBR$0"] = number
+    return query
+
+def run_query(query):
+    with requests.Session() as s:
+        new_form = BeautifulSoup(s.get(orig_link).text, 
+                                 "html.parser").find_all("input", type="hidden")
+        new_form = {x["id"]:x["value"] for x in new_form}
+        query.update({x:y for x, y in new_form.items() if y.strip() != ''})
+        query['ICAction'] = 'CLASS_SRCH_WRK2_SSR_PB_CLASS_SRCH'
+        r = s.post(orig_link, data=query, 
+                   headers={'Content-Type':"application/x-www-form-urlencoded"})
+    return r.text
+
+# Extracting Timetable
+def group_by_eq(seq, equalizer):
+    equiv_classes = {}
+    for elt in seq:
+        eq = equalizer(elt)
+        if eq not in equiv_classes:
+            equiv_classes[eq] = []
+        equiv_classes[eq].append(elt)
+    return equiv_classes
+
+def search_tag(tag, tag_name, attribute, string, matcher=(lambda x,y:
+                                                    re.search(x, y) is not None)):
+    try:
+        if re.compile(tag_name, re.I).match(tag.name):
+            return (tag.has_attr(attribute)
+                    and matcher(string, tag[attribute]))
+    except:
+        return False
+
+tag_is_course = lambda x: search_tag(x, "div", 
+                            "id", "win0divSSR_CLSRSLT_WRK_GROUPBOX2$", 
+                            lambda x,y: y.startswith(x))
+course_tag_is_title = lambda x: search_tag(x, "div", "id", 
+                            "win0divSSR_CLSRSLT_WRK_GROUPBOX2GP",
+                            lambda x,y: y.startswith(x))
+course_tag_is_section = lambda x: search_tag(x, "tr", "id", "trSSR_CLSRCH_MTG")
+section_tag_is_classname = lambda x: search_tag(x, "a", "id", "MTG_CLASSNAME")
+
+#@TODO Break into smaller subroutines
+def extract_timetables(string, year, term):
```

Comment:
I've said this before but I don't think long functions are that big of a deal (until you start doing automated tests) so for the moment I think this can stay the way it is (which to be fair I think is pretty much what was already gonna happen XD).

On another front however it would greatly increase readability if there were some comments in this function, at least to separate each part of the function (this might even help you figure out where to branch into different functions later).

Reply:
Agreed, that's definitely part of the plan!  Also included in the plan is differentiating public from private functions (at least to the extent python lets you do that), and adding docstrings and other such housekeeping.

In either case, as I may have mentioned elsewhere, I want to create a slightly more general framework that you can subclass to create scrapers for various universities to help improve extensibility down the line, so when that happens (hopefully soon!) I think that will solve a lot of these problems.  For this PR I was really just trying to focus on getting a minimum working thing going, so not sure how you want to handle that.

Also, on the topic of testing, it's something I've been thinking about a bit.  I do want to implement unit and integration tests (I may have this terminology wrong, so bear with me), but since this involves a lot of dealing with other peoples' messy, and more importantly, constantly changing, data, I haven't yet decided how I want to do that in a sustainable way.  Suggestions are more than welcome!

Reply:
I'm totally on board with all the readability changes you want to make, just keep in mind that the longer you put off doing that stuff the harder it gets to actually do later. I don't think it's an issue right now (this PR has a whole bunch of work in it as it is), just something to keep in mind.